### ISSUE TITLE
proposal: `Morphology`, `Sampler` and `PoseSet` models

### ISSUE THREAD

**Issue by alisterburt** [morphometrics/morphosamplers#21]
### Description

I'd like it if we could get to a place where there was some set of underlying models/interfaces around which this repo was based - heres a first pass at what those could look like

```python
class Morphology(BaseModel):
    """A set of attributes defining a geometrical support which can be sampled by a `Sampler`."""
    pass

class PoseSet(BaseModel):
    """Frozen model for a set of 3D poses."""
    positions: np.ndarray  # (n, 3)
    orientations: np.ndarray  # (n, 3, 3)

M = TypeVar("M", bound="Morphology")

class SamplerProtocol(Protocol[M]):
    """Protocol for classes that sample a morphology in a particular way."""
    @abstractmethod
    def __init__(self, **sampler_kwargs) -> None:
        ...

    @classmethod
    @abstractmethod
    def sample(self, obj: M) -> PoseSet:
        """Sample the morphology."""
        ...

SamplerType = TypeVar("SamplerType", bound=SamplerProtocol)
```

thoughts @kevinyamauchi @brisvag?

edit: code above has been updated after https://github.com/morphometrics/morphosamplers/issues/21#issuecomment-1481559208 to make it so model attributes should be set and sampler can be reused across morphologies

**Comment by brisvag** [comment]
So in practice you'll have some classes like:

```py
class SplineHelixSampler:
	def sample(obj: Spline3D, rise, twist, ...):
		# same as currently open PR for helical sampling

class SplineVolumeSampler:
	def sample(obj: Splin3D, tickness, spacing, ...):
 		# same as current behaviour for sampling volume along spline
```

and so on?

>     orientations: np.ndarray  # (n, 3, 3)

while I'm more and more convinced that scipy `Rotation`s are not a great object to juggle around (though they are still awesome for operations), I still don't love using matrices. It's weird to work with due to the extra dim, has redundant info, and needs to be checked because it can be an invalid rotation. Ideally I'd rather use quaternions, or even (*sigh*) euler angles...

**Comment by alisterburt** [comment]
Yep something like that - alternatively the rise, twist etc be model attributes and sampling only takes the Morphology

> I still don't love using matrices. It's weird to work with due to the extra dim, has redundant info, and needs to be checked because it can be an invalid rotation. Ideally I'd rather use quaternions, or even (sigh) euler angles...

I'm not sure what to suggest to be honest, in a sampling library the basis vectors will be the things that are used the most, those are the most accessible from the matrix and any other rep will force conversion between different representations...
- if we're constructing the matrices well there's no reason for the rotation to be invalid
- the only way you get around the extra dim is by hiding it in some transform object or einsumming directly
- I'm not sure what you mean by redundant info, it's just the basis vectors of the oriented thing 🤷🏼‍♂️

**Comment by alisterburt** [comment]
feedback from @kevinyamauchi - not all sampling operations will return images so that needs to be expanded, otherwise great!

**Comment by alisterburt** [comment]
Perhaps different sampler types - `PoseSampler`, `ImageSampler`, `LabelSampler` etc

**Comment by brisvag** [comment]
> * if we're constructing the matrices well there's no reason for the rotation to be invalid

Sure, and if we write code well there's no reason to have tests :P But yeah, this is not a strong issue for me, I just wanted to point it out.

> * the only way you get around the extra dim is by hiding it in some transform object or einsumming directly

Not following... Do you mean we could/should have a "transform" object ourselves? 
 
> * I'm not sure what you mean by redundant info, it's just the basis vectors of the oriented thing 🤷🏼‍♂️

I mean we're using 9 values to represent something that you can represent with 3 or 4. Memory/speed can be a concern, and this ties back into the invalid matrices too.

**Comment by alisterburt** [comment]
> Not following... Do you mean we could/should have a "transform" object ourselves?

Nah - I would say that's out of scope for this library, I just meant that the only way you get around having to change (n, 3) to (n, 3, 1) is by hiding it inside a transform object which does an explicit matmul or using einsum

**Comment by brisvag** [comment]
> If we end up in a situation where this becomes a concern let's reconsider but for the same cost as a 1k * 1k image (smol) we get 100k oriented points with the position + matrix rep

Fair enough!
